File: Problems & Solutions/Z/Zero Sum Subarrays/22 December 2022/Python/Zero Sum Subarrays.py
class Solution:
     def findSubArrays(self,arr,n):
      
        c=0
        hashMap = {} 
        out = [] 
        sum1 = 0
        
        for i in range(n):
            sum1 += arr[i] 
            
            if sum1 == 0: 
                out.append((0, i))
                c+=1
            al = [] 
            
            if sum1 in hashMap: 
                  
                al = hashMap.get(sum1) 
                for it in range(len(al)): 
                    out.append((al[it] + 1, i))
                    c+=1
                    
            al.append(i) 
            hashMap[sum1] = al 
                
        return c

#{ 
 # Driver Code Starts
#Initial Template for Python 3

#contributed by RavinderSinghPB
        
if __name__=='__main__':
    t=int(input())
    for tc in range(t):
        
        N=int(input())
        A = [int(x) for x in input().strip().split(' ')]
        ob = Solution()
        print(ob.findSubArrays(A,N))
        
                
                
# } Driver Code Ends

File: Problems & Solutions/Z/Zero Sum Subarrays/22 December 2022/JavaScript/Zero Sum Subarrays.js
//{ Driver Code Starts
//Initial Template for javascript
'use strict';

process.stdin.resume();
process.stdin.setEncoding('utf-8');

let inputString = '';
let currentLine = 0;

process.stdin.on('data', inputStdin => {
    inputString += inputStdin;
});

process.stdin.on('end', _ => {
    inputString = inputString.trim().split('\n').map(string => {
        return string.trim();
    });
    
    main();    
});

function readLine() {
    return inputString[currentLine++];
}

function main() {
    let t = parseInt(readLine());
    for(let i=0;i<t;i++)
    {
        let input_line = readLine().split(' ');
        let n = parseInt(input_line[0]);
        
        let arr = new Array(n);
        
        input_line = readLine().split(' ');
        for(let i=0;i<n;i++)
        {
            arr[i] = parseInt(input_line[i]);
        }
        
        let obj = new Solution();
        
        let ans = obj.findSubarray(arr, n)
        console.log(ans);
    }
}
// } Driver Code Ends

class Solution {
    findSubarray(arr, n){
        
        let mp = new Map();
        let sum = 0;
    	let counter=0;
    	
        for (let i = 0; i < n; i++)
    	{
    		sum = sum + arr[i];
    		
        	if (sum == 0)
    		{
    			counter++;
    		}
    		
    	    if(mp.has(sum)) 
            {
                counter+=mp.get(sum);
            }
    		
    		if(!mp.has(sum))
    		    mp.set(sum,1);
    		else
    		    mp.set(sum,mp.get(sum)+1);
    	}
    	return counter;   
    }
}